pydantic/main.py

```diff
@@ -428,10 +428,19 @@ def parse_obj(cls: Type['Model'], obj: Any) -> 'Model':
             try:
                 obj = dict(obj)
             except (TypeError, ValueError) as e:
-                exc = TypeError(f'{cls.__name__} expected dict not {obj.__class__.__name__}')
-                raise ValidationError([ErrorWrapper(exc, loc=ROOT_KEY)], cls) from e
+                exc_msg = f'{cls.__name__} expected dict not {obj.__class__.__name__}'
+                if isinstance(obj, list):
+                    exc_msg += ' (did you want to parse multiple objects with `many=True` ?)'
+                raise ValidationError([ErrorWrapper(TypeError(exc_msg), loc=ROOT_KEY)], cls) from e
         return cls(**obj)
 
+    @classmethod
+    def parse_obj(cls: Type['Model'], obj: Any, *, many: bool = False) -> Union['Model', List['Model']]:
```

Comment:
I decided to go for a `_parse_single_obj` for `mypy` instead of using `typing.overload` decorator as it introduced a regression on a test.
By writing something like this:
```python
@classmethod
@overload
def parse_obj(cls: Type['Model'], obj: Any, many: Literal[False]) -> 'Model': ...

@classmethod
@overload
def parse_obj(cls: Type['Model'], obj: Any, many: Literal[True]) -> List['Model']: ...

@classmethod
@overload
def parse_obj(cls: Type['Model'], obj: Any) -> 'Model':
    """The last overload is a fallback for mypy (https://mypy.readthedocs.io/en/stable/literal_types.html)"""

@classmethod
def parse_obj(cls: Type['Model'], obj: Any, many: bool = False) -> Union['Model', List['Model']]:  # noqa: F811
    if many:
        return [cls.parse_obj(x) for x in obj]
    ...
```

the test `tests/test_schema.py::test_literal_schema` started to fail because of the `Literal[1]` that didn't have the same `__values__`.

In fact it showed a bug that I didn't know:
```python
In [1]: from typing import Literal

In [2]: Literal[1]
Out[2]: typing.Literal[1]

In [3]: Literal[True]
Out[3]: typing.Literal[1]
```

```python
In [1]: from typing import Literal

In [2]: Literal[True]
Out[2]: typing.Literal[True]

In [3]: Literal[1]
Out[3]: typing.Literal[True]
```

So the annotations that were just added had in fact some side effect ! (WTF...)
As this bug is directly in python, I decided to just forget the `@overload` stuff and just go with this one instead

Reply:
@samuelcolvin As a matter of fact I think `parse_obj_as` is probably the best. I let the PR open but don't hesitate to close it as I'm not convinced it's pushing pydantic in the right direction on this topic